File: Exercism/javascript/twelve-days/twelve-days.js
'use strict'

const NUMBER = new Map([
  { word: 'zero', th: 'zeroth' },
  { word: 'and', th: 'first' },
  { word: 'two', th: 'second' },
  { word: 'three', th: 'third' },
  { word: 'four', th: 'fourth' },
  { word: 'five', th: 'fifth' },
  { word: 'six', th: 'sixth' },
  { word: 'seven', th: 'seventh' },
  { word: 'eight', th: 'eighth' },
  { word: 'nine', th: 'ninth' },
  { word: 'ten', th: 'tenth' },
  { word: 'eleven', th: 'eleventh' },
  { word: 'twelve', th: 'twelfth' },
].entries())

const PRESENTS = [
  'Drummers Drumming',
  'Pipers Piping',
  'Lords-a-Leaping',
  'Ladies Dancing',
  'Maids-a-Milking',
  'Swans-a-Swimming',
  'Geese-a-Laying',
  'Gold Rings',
  'Calling Birds',
  'French Hens',
  'Turtle Doves',
  'a Partridge in a Pear Tree.'
]

export const recite = (from, to = undefined) => {
  let verse = ''
  to = to || from
  for (let i = from; i <= to; i += 1) {
    const presents = PRESENTS
      .slice(PRESENTS.length - i)
      .map((p, idx, arr) => `${
        arr.length > 1
          ? ' ' + NUMBER.get(arr.length - idx).word
          : ''
        } ${p}`)
      .join(',')
    verse += `On the ${NUMBER.get(i).th} day of Christmas my true love gave to me:${presents}\n\n`
  }
  return verse.replace(/\n$/, '')
}

File: Exercism/python/twelve-days/twelve_days.py
PRESENTS = [
    'a Partridge in a Pear Tree',
    'two Turtle Doves',
    'three French Hens',
    'four Calling Birds',
    'five Gold Rings',
    'six Geese-a-Laying',
    'seven Swans-a-Swimming',
    'eight Maids-a-Milking',
    'nine Ladies Dancing',
    'ten Lords-a-Leaping',
    'eleven Pipers Piping',
    'twelve Drummers Drumming'
]

NUMBERS = [
    'first',
    'second',
    'third',
    'fourth',
    'fifth',
    'sixth',
    'seventh',
    'eighth',
    'ninth',
    'tenth',
    'eleventh',
    'twelfth'
]


def recite(start_verse: int, end_verse: int) -> list:
    """Outputs the lyrics to 'The Twelve Days of Christmas'."""
    verse = []
    for i in range(start_verse, end_verse + 1):
        verse.append(f'On the {NUMBERS[i-1]} day of Christmas'
                     ' my true love gave to me: '
                     f'{", ".join(PRESENTS[i-1:0:-1])}'
                     f'{", and " if i > 1 else ""}'
                     f'{PRESENTS[0]}.')
    return verse
